### PR TITLE
Add Uml2Java move classifier reaction

### PR DESCRIPTION
**Pull request by JanWittler**
Currently a move uml classifier operation is not propagated to Java at all as there is no reaction specified for it.
This pull request adds a reaction to the Uml2Java reactions to react to a UML classifier move and adds test cases for uml class and uml data type move.

### Discussion

The ```VitruviusChange``` representation of a _Move_ operation is a ```RemoveEReference``` and an ```InsertEReference``` change. However, the proposed reaction does only react to the _insert_ change due to following reasons:
- A stand-alone ```RemoveEReference``` change is not possible as the parts of the UML document have to either be moved to another location or otherwise are not present in the document anymore and are thus deleted. As there is already a reaction to _deleted and removed_, one can assume that a _deleted without removed_ will be followed by an _insert_.
- Handling the ```RemoveEReference``` directly will produce incorrect results
  - Deleting the ```CompilationUnit``` deletes the class, thus it will not be available to the subsequent ```InsertEReference``` anymore.
  - Removing the ```CompilationUnit``` from the package will produce a model in an invalid state as the ```CompilationUnit``` is still stored at the old location (as there is no location to store it correctly).
- The Reactions language does not support declaring reactions to composite changes.

### CODE REVIEW

**Comment by jGleitz** [comment]
@HeikoKlare we can already match certain change sequences in the reactions language, like “created and inserted at”. Is there no way to match “removed from and inserted at”? And if there isn’t, wouldn’t this be a logical, low-cost addition?

**Comment by HeikoKlare** [comment]
Currently, there is no "removed from and inserted at" sequence that the reactions language is able to match. We can, of course, add a syntax extension and sequence matcher for that. We had some long discussions about whether matching change sequences within reaections is reasonable or not. The currently available sequence of existence and feature changes are not used properly in most of the cases and we had actually concluded that are unlikely to be reasonable at all. That is why I am not sure whether having a move trigger is good or necessary. I do actually prefer to find out whether an element insertion is the consequence of a relevant move operation by context information, such as the corresponding elements (just like Jan does in his changes). The reason for that is that you are actually not interested in whether an element was moved from some specific location (i.e., a specific reference), but whether an element is inserted that already instantiated a specific mapping (which you do by looking at the correspondences). Imagine that you may have a classifier referenced from a compilation unit or from another classifier (so it is a nested classifier). Then whenever you move the classifier somewhere else, you are not interested in whether it was contained within a compilation unit or another classifier before, but you are want to know whether an according mapping of that classifier has already been instantiated that you can reuse.

**Comment by jGleitz** [comment]
I see. Sounds reasonable.

However, I think that by the same token, the ‘created and inserted at’ syntax & matcher should be removed as well.

**Comment by HeikoKlare** [comment]
Fair point. I have added an issue to document the necessity to clean up the reactions to be able to remove the language feature afterwards.

File: bundles/tools.vitruv.applications.umljava/src/tools/vitruv/applications/umljava/uml2java/UmlToJavaClassifier.reactions [hunk 1]
```diff
@@ -106,6 +106,36 @@ routine createJavaCompilationUnit(uml::Classifier umlClassifier, java::ConcreteC
...
+		val javaCompilationUnit = retrieve java::CompilationUnit corresponding to umlClassifier 
+			with !javaPackage.isPresent || !javaPackage.get.compilationUnits.contains(javaCompilationUnit)
+	}
+	action {
+		update javaCompilationUnit {
+			javaCompilationUnit.namespaces.clear
+			javaPackage.ifPresent [
+				javaCompilationUnit.namespaces += javaPackageAsStringList
+			]
+			//TODO: updating the compilation unit's name should not be needed
```
Comment (tsaglam) [inline bundles/tools.vitruv.applications.umljava/src/tools/vitruv/applications/umljava/uml2java/UmlToJavaClassifier.reactions:127]: I checked how this is handled in other routines, and it seems to be the standard. For example here:
https://github.com/vitruv-tools/Vitruv-Applications-ComponentBasedSystems/blob/97ddb33878267e365b8f704eb93792bf673d7f70/bundles/tools.vitruv.applications.umljava/src/tools/vitruv/applications/umljava/uml2java/UmlToJavaClassifier.reactions#L129
Thus I think we can remove the todo-comment.

**Review by tsaglam — CHANGES_REQUESTED** [review]

File: bundles/tools.vitruv.applications.umljava/src/tools/vitruv/applications/umljava/uml2java/UmlToJavaClassifier.reactions (hunk 1 above)
Comment (JanWittler) [reply bundles/tools.vitruv.applications.umljava/src/tools/vitruv/applications/umljava/uml2java/UmlToJavaClassifier.reactions:127]: I think the referenced example is chose a bit unluckily as it's a rename operation and rename operations obviously need to update the name.
I searched the PCM -> Java and the other UML -> Java reactions for an actual move reaction as reference but did not find any so I don't know if there even is any _standard behaviour_ for a ```move``` operation currently.

Comment (tsaglam) [reply bundles/tools.vitruv.applications.umljava/src/tools/vitruv/applications/umljava/uml2java/UmlToJavaClassifier.reactions:127]: Good point. Then let us keep the comment for now as we can always decide to remove it.

**Review by tsaglam — APPROVED** [review]

**Pull request merged**